_drafts/firmware-code-size-optimization-1.md

```diff
@@ -0,0 +1,302 @@
+---
+title: "Code size: first, measure it!"
+author: francois
+---
+
+<!-- excerpt start -->
+
+Every firmware engineer has run out of code space at some point or another.
+Whether they are trying to cram in another feature, or to make enough space for
+[A/B firmware
+updates](https://sbabic.github.io/swupdate/overview.html#double-copy-with-fall-back)
+more code space is always better.
+
+In this series of posts, we'll explore ways to save code space and ways not to
+do it. We will cover compiler options, coding style, logging, as well as
+desperate hacks when all you need is another 24 bytes.
+
+But first, let's talk about measuring code size.
+
+<!-- excerpt end -->
+
+### Why measure?
+
+Optimization is often counter intuitive. Don't take my word for it: this is one
+of the topics [Microsoft's Raymond
+Wong](https://devblogs.microsoft.com/oldnewthing/20041216-00/?p=36973) and [Unix
+programmers](https://homepage.cs.uri.edu/~thenry/resources/unix_art/ch12s02.html)
+agree on!
+
+So before you do anything, measure where your code size is going. You may be
+surprised!
+
+### Measuring overall code size
+
+The simplest way to measure code size is to inspect the different sections of
```

Comment:
`-Wl,--print-memory-usage` is pretty cool too. Suffers the fate of data being lumped into ram count but at link you get something like:

```
Memory region         Used Size  Region Size  %age Used
             rom:       10800 B       256 KB      4.12%
             ram:        8376 B        32 KB     25.56%
```

Reply:
Added a note, though this is an inferior solution to `size` because you cannot compute your actual ROM usage from this output, since it does not break out data vs. bss